src/cargo/util/toml/mod.rs

```diff
@@ -183,10 +184,33 @@ fn read_manifest_from_str(
             }
         }
     }
+
+    let mut public_deps_without_z = BTreeSet::new();
+    if let Some(deps) = &mut manifest.dependencies {
+        for (name, dep) in deps {
+            if let InheritableDependency::Value(toml_dep) = dep {
+                if toml_dep.is_public() && !config.cli_unstable().public_dependency {
+                    if let TomlDependency::Detailed(d) = toml_dep {
+                        d.public = Some(false)
+                    }
+                    public_deps_without_z.insert(name.clone());
+                }
+            }
+        }
+    }
+    let public_warn = |warnings: &mut Warnings| {
+        for name in public_deps_without_z {
+            warnings.add_warning(format!(
+                "{name} is public, pass `-Zpublic-dependency` to enable support for it",
+            ))
+        }
```

Comment:
Why is this split into two separate passes, rather than adding the warnings while we make the changes?

Reply:
The `public_warn`  is a closure function which need a `Manifest.warnings` as a input param.
The `public_warn` will invoke in latter, because the `Manifest.warnings` was produced by `to_real_manifest` or `to_virtual_manifest`.

Besides,  may be this need a `BTreeSet` to deduplicate.

The `public_warn` is similar to `add_unused`, I can combine them into one.

Reply:
> I had migrated the code to here, because detailed_dep_to_dependency will invoke by read_manifest_from_str.

These code had deleted.

Reply:
> Why is this split into two separate passes, rather than adding the warnings while we make the changes?

This is similar to `add_unused`.  It can't warning in the same passes, because we can't get `Warning` instance now.